packages/doenetml-worker-rust/lib-doenetml-core/src/core/state_var_interfaces/text_state_var_interfaces.rs

```diff
@@ -0,0 +1,197 @@
+                self.from_single_essential = true;
+            }
+        }
+    }
+
+    fn calculate_state_var_from_dependencies_and_mark_fresh(
+        &self,
+        state_var: &StateVarMutableViewTyped<String>,
+    ) {
+        if self.from_single_essential {
+            // if we are basing it on a single essential variable,
+            // then we propagate used_default as well as the value.
+            state_var.set_value_and_used_default(
+                self.string_dependency_values[0].get_fresh_value().clone(),
+                self.string_dependency_values[0].get_used_default(),
+            );
+        } else {
+            // TODO: can we implement this without cloning the inner value?
+            let value: String = self
+                .string_dependency_values
+                .iter()
+                .map(|v| v.get_fresh_value().clone())
+                .collect();
+
+            state_var.set_value(value);
+        }
+    }
+
+    fn request_dependencies_to_update_value(
+        &self,
+        state_var: &StateVarReadOnlyViewTyped<String>,
+        _is_direct_change_from_renderer: bool,
+    ) -> Result<Vec<DependencyValueUpdateRequest>, RequestDependencyUpdateError> {
+        if self.string_dependency_values.len() != 1 {
+            // TODO: implement for no dependencies where saves to essential value?
+            Err(RequestDependencyUpdateError::CouldNotUpdate)
+        } else {
+            let desired_value = state_var.get_requested_value();
+
+            self.string_dependency_values[0].request_change_value_to(desired_value.clone());
+
+            Ok(vec![DependencyValueUpdateRequest {
+                instruction_idx: 0,
+                dependency_idx: 0,
+            }])
+        }
+    }
+}
+
+/// A simplified version of GeneralStringStateVarInterface
+/// that is based on a single dependency.
+/// Requires a `dependency_instruction_hint`.
+///
+/// Panics if `create_dependency_from_extend_source` is true,
+/// if not given a `dependency_instruction_hint`,
+/// or if given a `dependency_instruction_hint`
+/// that doesn't result in at least one string dependency.
+#[derive(Debug, Default)]
+pub struct SingleDependencyStringStateVarInterface {
+    string_dependency_value: StateVarReadOnlyViewTyped<String>,
+}
+
+impl StateVarInterface<String> for SingleDependencyStringStateVarInterface {
+    fn return_dependency_instructions(
+        &self,
+        _extend_source: Option<&ExtendSource>,
+        parameters: &StateVarParameters,
+    ) -> Vec<DependencyInstruction> {
+        if parameters.create_dependency_from_extend_source {
+            panic!("Cannot create dependency from extend source for SingleDependencyStringStateVarInterface");
+        }
+
+        let mut dep_instructs: Vec<DependencyInstruction> = Vec::with_capacity(1);
+
+        if let Some(dependency_instruction) = &parameters.dependency_instruction_hint {
+            dep_instructs.push(dependency_instruction.clone());
```

Comment:
This function is getting bigger? Is that why you wrote it this way? Otherwise you could return a `vec![...]` right here.

Reply:
No, the function was getting smaller, as I copied it from `GeneralStringStateVarInterface` and didn't take advantage of the new simplicity.